src/ctapipe/monitoring/calculator.py

```diff
@@ -0,0 +1,327 @@
+            col_name=col_name,
+            chunk_shift=None,
+        )
+        # Detect faulty pixels with multiple instances of ``OutlierDetector``
+        outlier_mask = np.zeros_like(aggregated_stats["mean"], dtype=bool)
+        for aggregated_val, outlier_detector in self.outlier_detectors.items():
+            outlier_mask = np.logical_or(
+                outlier_mask,
+                outlier_detector(aggregated_stats[aggregated_val]),
+            )
+        # Add the outlier mask to the aggregated statistics
+        aggregated_stats["outlier_mask"] = outlier_mask
+        # Get valid chunks and add them to the aggregated statistics
+        aggregated_stats["is_valid"] = self._get_valid_chunks(outlier_mask)
+        return aggregated_stats
+
+    def second_pass(
+        self,
+        table,
+        valid_chunks,
+        tel_id,
+        masked_pixels_of_sample=None,
+        col_name="image",
+    ) -> Table:
+        """
+        Conduct a second pass over the data to refine the statistics in regions with a high percentage of faulty pixels.
+
+        This method performs a second pass over the data with a refined shift of the chunk in regions where a high percentage
+        of faulty pixels were detected during the first pass. Note: Multiple first passes of different calibration events are
+        performed which may lead to different identification of faulty chunks in rare cases. Therefore a joined list of faulty
+        chunks is recommended to be passed to the second pass(es) if those different passes use the same ``chunk_size``.
+
+        Parameters
+        ----------
+        table : astropy.table.Table
+            DL1-like table with images of shape (n_images, n_channels, n_pix), event IDs and timestamps of shape (n_images, ).
+        valid_chunks : ndarray
+            Boolean array indicating the validity of each chunk from the first pass.
+            Note: This boolean array can be a ``logical_and`` from multiple first passes of different calibration events.
+        tel_id : int
+            Telescope ID for which the calibration is being performed.
+        masked_pixels_of_sample : ndarray, optional
+            Boolean array of masked pixels of shape (n_pix, ) that are not available for processing.
+        col_name : str
+            Column name in the table from which the statistics will be aggregated.
+
+        Returns
+        -------
+        astropy.table.Table
+            Table containing the aggregated statistics after the second pass, their outlier masks, and the validity of the chunks.
+        """
+        # Check if the chunk_shift is set for the second pass
+        if self.chunk_shift is None:
+            raise ValueError(
+                "chunk_shift must be set if second pass over the data is requested"
+            )
+        # Check if at least one chunk is faulty
+        if np.all(valid_chunks):
+            raise ValueError(
```

Comment:
Is this really an error? Or should it just return as a no-op?

Reply:
What I missed until now: there are only `first_pass` and `second_pass` methods here, but now common one.

What is the intendend use? Can we add a `__call__` along the lines of:

```
def __call__(self, ...):
    result_first = self.first_pass(...)
    second_pass_necessary = ...
    if second_necessary_necessary:
        result_second = self.second_pass(...)
    # merge results
    result = ...
    return result
```
?

Reply:
This is currently done in the tool, i.e. see [here](https://gitlab.cta-observatory.org/cta-computing/dpps/calibrationpipeline/calibpipe/-/blob/fcd10f4ff3f04f5e01af4d77312760d4250a448c/src/calibpipe/tools/stats_calculation.py#L124-151). Should this be moved into the `PixelStatisticsCalculator` component?

Reply:
Yes, I think this is purely concerning this extractor here.

Reply:
And without it, the intended use is not obvious, the code is incomplete

Reply:
While I agree that as is the intended use is not obvious, I'm not sure we want to implement the logic of first/second pass calling here. There might be situations when you don't want to call the second pass (e.g. first pass detected too many problematic chunks and you conclude your data is basically trash). I suggest to add to the class docstring the example of use like @maxnoe pointed above, but leave the decisive logic to the tool that calls the component.

Reply:
We discussed the boundary of ctapipe / calibpipe in a meeting and decided that the statistics calculators are useful to have in ctapipe.

A generic tool to apply these to data would also be good as a follow up PR. The rest stays / goes into calibpipe